apps/hash-ai-worker-ts/src/activities/flow-activities/research-entities-action/link-follower-agent/get-link-follower-next-tool-calls.ts

```diff
@@ -0,0 +1,317 @@
+  </TaskDescription>
+`);
+
+type GetLinkFollowerNextToolCallsParams = {
+  task: string;
+  entitySummaries: LocalEntitySummary[];
+  factsGathered: Fact[];
+  previouslyVisitedLinks: { url: string }[];
+  possibleNextLinks: Link[];
+};
+
+const simplifyFactForLlmConsumption = (fact: Fact) => {
+  return `${fact.text} ${fact.prepositionalPhrases.join(", ")}`;
+};
+
+const generateUserMessage = (
+  params: GetLinkFollowerNextToolCallsParams,
+): LlmUserMessage => {
+  const {
+    task,
+    entitySummaries,
+    factsGathered,
+    previouslyVisitedLinks,
+    possibleNextLinks,
+  } = params;
+
+  return {
+    role: "user",
+    content: [
+      {
+        type: "text",
+        text: dedent(`
+Task: ${task}
+Previously Visited Links: ${JSON.stringify(previouslyVisitedLinks)}
+Entities: ${JSON.stringify(
+          entitySummaries.map(({ localId, name, summary, entityTypeId }) => {
+            const factsAboutEntity = factsGathered.filter(
+              (fact) => fact.subjectEntityLocalId === localId,
+            );
+
+            return {
+              name,
+              summary,
+              entityType: entityTypeId,
+              facts: factsAboutEntity,
+            };
+          }),
+        )}
+Facts Gathered: ${JSON.stringify(factsGathered.map(simplifyFactForLlmConsumption))}
+Possible Next Links: ${JSON.stringify(possibleNextLinks)}
+    `),
+      },
+    ],
+  };
+};
+
+const toolNames = ["exploreLinks", "complete", "terminate"] as const;
+
+type ToolName = (typeof toolNames)[number];
+
+const suggestionForNextStepsDefinition = {
+  type: "string",
+  description: dedent(`
+    A suggestion for how to find any relevant facts that could be used to provide values for additional properties.
+    This should be a detailed explanation of how you would go about finding the missing facts from online resources.
+    If the you've encountered URLs for web pages which may be relevant, you must include them in the suggestion.
+  `),
+};
+
+const tools: LlmToolDefinition<ToolName>[] = [
+  {
+    name: "exploreLinks",
+    description: dedent(`
+      Explore the links specified to gather more facts and fulfill the task.
+      The links must be one of the possible next links provided by the user.
+    `),
+    inputSchema: {
+      type: "object",
+      properties: {
+        links: {
+          type: "array",
+          items: {
+            type: "object",
+            properties: {
+              url: {
+                type: "string",
+                description: "The URL of the link to explore",
+              },
+              reason: {
+                type: "string",
+                description: "The reason for exploring this link",
+              },
+              descriptionOfExpectedContent: {
+                type: "string",
+                description: dedent(`
+                  A description of the content you expect to find at the link.
```

Comment:
I'm not sure it's 'expect' so much as 'hope' or 'would like', but don't know if this makes any difference in practice so I wouldn't change it